base/codevalidation.jl

```diff
@@ -0,0 +1,159 @@
+                        push!(errors, InvalidCodeError(3, "encountered invalid call argument", arg))
+                    end
+                end
+            end
+        elseif isa(x, SSAValue) && !in(x, ssavals)
+            push!(ssavals, x)
+        end
+    end
+    if isempty(c.slotnames)
+        push!(errors, InvalidCodeError(4, "slotnames field is empty"))
+    end
+    if length(c.slotnames) != length(c.slotflags)
+        msg = "length(slotflags) != length(slotnames)"
+        push!(errors, InvalidCodeError(5, msg, (length(c.slotflags), length(c.slotnames))))
+    end
+    if c.inferred
+        if length(c.slottypes) != length(c.slotnames)
+            msg = "length(slottypes) != length(slotnames)"
+            push!(errors, InvalidCodeError(6, msg, (length(c.slottypes), length(c.slotnames))))
+        end
+        if length(c.ssavaluetypes) < length(ssavals)
+            missing = length(ssavals) - length(c.ssavaluetypes)
+            msg = "not all SSAValues in AST have a type in ssavaluetypes"
+            push!(errors, InvalidCodeError(7, msg, missing))
+        end
+    else
+        if c.slottypes !== nothing
+            msg = "uninferred CodeInfo slottypes field is not `nothing`"
+            push!(errors, InvalidCodeError(8, msg, c.slottypes))
+        end
+        if c.ssavaluetypes != length(ssavals)
+            msg = "uninferred CodeInfo ssavaluetypes field does not equal the number of present SSAValues"
+            push!(errors, InvalidCodeError(9, msg, (length(ssavals), c.ssavaluetypes)))
+        end
+    end
+    checked_ids = Int[]
+    walkast(c.code) do x
+        if isa(x, Expr) && x.head == :(=)
+            lhs = x.args[1]
+            if isa(lhs, SlotNumber) && !is_flag_set(c.slotflags[lhs.id], ASSIGNED_FLAG) && !in(lhs.id, checked_ids)
+                msg = "slot has wrong assignment slotflag setting (bit flag 2 should be set)"
+                push!(errors, InvalidCodeError(10, msg, lhs.id))
+                push!(checked_ids, lhs.id)
+            end
+        end
+    end
+    return errors
+end
+
+"""
+    validate_code!(errors::Vector{>:InvalidCodeError}, m::Method)
+
+Validates `m`, logging any violations in `errors`.
+
+After all `Method` checks are complete, `validate_code!(errors, uncompress_ast(m))` is called.
+
+Any violation of the following properties will be logged by pushing a `InvalidCodeError`
+into `errors` (where `c = uncompress_ast(m)`):
+
+- `length(m.sig.parameters) == m.nargs`
+- `m.isva == isa(last(m.sig.parameters), Vararg)`
+- `m.nargs <= length(c.slotnames)`
+- `h != :method` for any subexpression head `h` if `m.nargs > 0`
+"""
+function validate_code!(errors::Vector{>:InvalidCodeError}, m::Method)
+    if length(m.sig.parameters) != m.nargs
+        msg = "number of types in method signature does not match number of arguments"
+        push!(errors, InvalidCodeError(11, msg, (length(m.sig.parameters), m.nargs)))
+    end
+    if m.isva != (last(m.sig.parameters) <: Vararg{Any})
```

Comment:
These two tests aren't quite correct. If `isva` is true, at most you can say about `length(unwrap_unionall(m.sig.parameters))` is that it is `>= m.nargs - 1`. Otherwise, it is true that they must be exactly equal.